File: app/src/main/java/com/artemchep/horario/_new/fragments/UserEditFragment.java
package com.artemchep.horario._new.fragments;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.artemchep.horario.R;
import com.artemchep.horario.database.Persy;
import com.artemchep.horario.database.models.User;
import com.artemchep.horario.ui.UiHelper;
import com.artemchep.horario.ui.activities.ActivityHorario;
import com.bumptech.glide.Glide;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.stfalcon.frescoimageviewer.ImageViewer;

import java.io.IOException;

import de.hdodenhof.circleimageview.CircleImageView;

/**
 * @author Artem Chepurnoy
 */
public class UserEditFragment extends FragmentBase implements View.OnClickListener, Toolbar.OnMenuItemClickListener {

    public static final String EXTRA_USER = "extra::user";
    public static final String EXTRA_USER_ID = "extra::user::id";

    private static final int PICK_IMAGE_REQUEST = 228;

    private static String[] ANIMALS = new String[]{
            "Alligator", "Anteater", "Armadillo", "Auroch",
            "Axolotl", "Badger", "Bat", "Beaver",
            "Buffalo", "Camel", "Chameleon", "Cheetah",
            "Chipmunk", "Chinchilla", "Chupacabra", "Cormorant",
            "Coyote", "Crow", "Dingo", "Dinosaur",
            "Dolphin", "Duck", "Elephant", "Ferret",
            "Fox", "Frog", "Giraffe", "Gopher",
            "Grizzly", "Hedgehog", "Hippo", "Hyena",
            "Jackal", "Ibex", "Ifrit", "Iguana",
            "Koala", "Kraken", "Lemur", "Leopard",
            "Liger", "Llama", "Manatee", "Mink",
            "Monkey", "Narwhal", "Nyan Cat", "Orangutan",
            "Otter", "Panda", "Penguin", "Platypus",
            "Python", "Pumpkin", "Guagga", "Rabbit",
            "Raccoon", "Rhino", "Sheep", "Shrew",
            "Skunk", "Slow Loris", "Squirrel", "Turtle",
            "Walrus", "Wolf", "Wolverine", "Wombat"
    };

    private User mUser;
    private String mUserId;

    private Toolbar mToolbar;
    private ImageView mUserAvatarImageView;
    private TextView mUserNameTextView;
    private EditText mInfoEmailEditView;
    private EditText mInfoUserNameEditView;
    private EditText mInfoUserDetailsEditView;

    private Persy.Watcher<User> mWatcher;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Bundle args = getArguments();
        mUser = args.getParcelable(EXTRA_USER);
        mUser = mUser.clone();
        mUserId = args.getString(EXTRA_USER_ID, "");

        ActivityHorario activity = (ActivityHorario) getActivity();
        mWatcher = activity.getPersy().watchFor(User.class, "users/" + mUserId);
    }

    @Nullable
    @Override
    public View onCreateView(
            LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.___fragment_user_edit, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mToolbar = view.findViewById(R.id.toolbar);
        mToolbar.inflateMenu(R.menu.user_edit);
        mToolbar.setOnMenuItemClickListener(this);
        mUserAvatarImageView = (CircleImageView) view.findViewById(R.id.user_avatar);
        mUserAvatarImageView.setOnClickListener(this);

        mInfoEmailEditView = view.findViewById(R.id.info_email);
        mInfoUserNameEditView = view.findViewById(R.id.info_user_name);
        mInfoUserDetailsEditView = view.findViewById(R.id.info_user_details);

        bind(mUser);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case PICK_IMAGE_REQUEST: {
                break;
            }
        }
    }

    private void bind(@Nullable User user) {
        if (user == null) {
            mUserAvatarImageView.setImageDrawable(null);
            return;
        }

        mInfoEmailEditView.setText(user.email);
        mInfoUserNameEditView.setText(user.name);
        mInfoUserDetailsEditView.setText(user.info);

        // Load avatar
        Glide.with(this)
                .load(user.avatarUrl)
                .fallback(R.drawable.ic_pencil_grey600_24dp)
                .into(mUserAvatarImageView);
    }

    /**
     * Generate Google Docs like name (Anonymous Nyan Cat, etc.)
     * for given key.
     */
    @NonNull
    private String generateNameFor(@NonNull String key) {
        return String.format("Anonymous %s", ANIMALS[key.hashCode() % ANIMALS.length]);
    }

    private void selectAvatar() {
        Intent intent = new Intent();
        intent.setType("image/*");
        intent.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(Intent.createChooser(intent, "Select Picture"), PICK_IMAGE_REQUEST);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.fab: {
                break;
            }
            case R.id.user_avatar: {
                // TODO: https://github.com/chrisbanes/PhotoView
                //noinspection unchecked
                new ImageViewer.Builder(getContext(), new String[]{
                        "https://habrastorage.org/getpro/tmtm/pictures/6cd/cc9/763/6cdcc97634f82f7c12590bbd2e56ff14.jpg"
                }).show();
                break;
            }
        }
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_save:
                updateCurrentUser();
                mWatcher.getDatabase().setValue(mUser);
                finish();
                break;
            default:
                return false;
        }
        return true;
    }

    private void updateCurrentUser() {
        mUser.email = mInfoEmailEditView.getText().toString();
        mUser.name = mInfoUserNameEditView.getText().toString();
        mUser.info = mInfoUserDetailsEditView.getText().toString();
    }

}

File: app/src/main/java/com/artemchep/horario/ui/fragments/UserEditFragment.kt
package com.artemchep.horario.ui.fragments

import android.content.Context
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.Toolbar
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import com.artemchep.horario.R
import com.artemchep.horario.database.data.User
import com.artemchep.horario.database.models.IUser
import com.artemchep.horario.interfaces.FragmentHost
import com.artemchep.horario.ui.utils.EXTRA_USER
import com.artemchep.horario.ui.utils.EXTRA_USER_ID
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import timber.log.Timber


/**
 * @author Artem Chepurnoy
 */
class UserEditFragment : FragmentBase(), View.OnClickListener {

    companion object {
        const val TAG = "UserEditFragment"
        const val STATE_USER = "state::user"
    }

    private lateinit var userId: String
    private lateinit var origin: User
    private lateinit var user: User

    private lateinit var fab: FloatingActionButton
    private lateinit var toolbar: Toolbar
    private lateinit var nameEditView: EditText
    private lateinit var biographyEditView: EditText
    private lateinit var emailEditView: EditText

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        arguments!!.apply {
            userId = getString(EXTRA_USER_ID)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        origin = arguments!!.getParcelable<IUser>(EXTRA_USER)?.let {
            User(
                    id = it.id,
                    name = it.name,
                    email = it.email,
                    biography = it.biography,
                    avatarUrl = it.avatarUrl
            )
        } ?: User()

        user = if (savedInstanceState != null) {
            savedInstanceState.getParcelable(STATE_USER)
        } else origin.copy()
    }

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.___fragment_user_edit, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        fab = view.findViewById<FloatingActionButton>(R.id.fab).apply {
            setOnClickListener(this@UserEditFragment)
        }
        toolbar = view.findViewById<Toolbar>(R.id.toolbar).apply {
            setNavigationOnClickListener { _ ->
                val host = activity as FragmentHost
                host.fragmentFinish(this@UserEditFragment)
            }
        }
        biographyEditView = view.findViewById(R.id.user_biography)
        emailEditView = view.findViewById(R.id.email)
        nameEditView = view.findViewById<EditText>(R.id.user_name).apply {
            addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}

                override fun afterTextChanged(editable: Editable) {
                    if (nameEditView.text.isNullOrEmpty()) {
                        fab.hide()
                    } else fab.show()
                }
            })
        }

        nameEditView.setText(user.name)
        emailEditView.setText(user.email)
        biographyEditView.setText(user.biography)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        updateCurrentUser()
        outState.putParcelable(STATE_USER, user)
    }

    private fun updateCurrentUser() {
        user.run {
            name = nameEditView.text.toString()
            email = emailEditView.text.toString()
            biography = biographyEditView.text.toString()
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.fab -> {
                updateCurrentUser()

                val success = dbSave()
                if (success) {
                    val host = activity as FragmentHost
                    host.fragmentFinish(this)
                }
            }
        }
    }

    private fun dbSave(): Boolean {
        val process: String?.() -> String? = {
            this?.trim()
                    ?.replace("\n", "")
                    ?.replace("\r", "")
        }

        user.name = user.name?.process()
        user.email = user.email?.process()
        user.biography = user.biography?.trim()

        nameEditView.setText(user.name)

        // Client-side data rules
        if (user.name.isNullOrBlank()) {
            return false
        }

        val firestore = FirebaseFirestore.getInstance()
        val doc = firestore.document("users/$userId")

        val curr: MutableMap<String, Any?> = user.deflateUser()
        val prev: MutableMap<String, Any?> = origin.deflateUser()

        // Update only changed entries
        val map = curr.filter { prev[it.key] != it.value }
        map.takeUnless { map.isEmpty() }?.also {
            Timber.tag(TAG).i("User MERGE: [$map]")
            doc.set(it, SetOptions.merge())
        }
        return true
    }

}
